sanic_openapi/swagger.py

```diff
@@ -228,46 +194,101 @@ def build_spec(app, loop):
-        for tag in route_spec.tags:
-            tags[tag] = True
-    _spec.add_tags(tags=[{"name": name} for name in tags.keys()])
+    @app.listener("after_server_start")
+    def build_spec(app, loop):
+        _spec = Spec(app=app)
+
+        # --------------------------------------------------------------- #
+        # Blueprint Tags
+        # --------------------------------------------------------------- #
+
+        for blueprint in app.blueprints.values():
+            build_blueprint(blueprint=blueprint)
+
+        paths = build_paths(app=app, url_prefix=url_prefix)
+
+        # --------------------------------------------------------------- #
+        # Definitions
+        # --------------------------------------------------------------- #
+
+        _spec.add_definitions(
+            definitions={
+                obj.object_name: definition for obj, definition in definitions.values()
+            }
+        )
+
+        # --------------------------------------------------------------- #
+        # Tags
+        # --------------------------------------------------------------- #
+
+        # TODO: figure out how to get descriptions in these
+        tags = {}
+        for route_spec in route_specs.values():
+            if route_spec.blueprint and route_spec.blueprint.name in ("swagger"):
+                # TODO: add static flag in sanic routes
+                continue
+            for tag in route_spec.tags:
+                tags[tag] = True
+        _spec.add_tags(tags=[{"name": name} for name in tags.keys()])
+
+        _spec.add_paths(paths)
+        app._spec = _spec
 
-    _spec.add_paths(paths)
-    swagger_blueprint._spec = _spec
+    @app.route(url_prefix + "/swagger.json")
+    @doc_route(exclude=True)
+    def _spec(request):
+        return response.json(app._spec)
 
+    @app.route(url_prefix + "/swagger-config")
+    def config(request):
+        options = {}
 
-@swagger_blueprint.route("/swagger.json")
-@doc_route(exclude=True)
-def spec(request):
-    return json(swagger_blueprint._spec)
+        if hasattr(request.app.config, "SWAGGER_UI_CONFIGURATION"):
+            options = getattr(request.app.config, "SWAGGER_UI_CONFIGURATION")
 
+        return response.json(options)
 
-@swagger_blueprint.route("/swagger-config")
-def config(request):
-    options = {}
 
-    if hasattr(request.app.config, "SWAGGER_UI_CONFIGURATION"):
-        options = getattr(request.app.config, "SWAGGER_UI_CONFIGURATION")
+class Swagger:
+    def __init__(self, app=None, url_prefix=None, spec=None):
+        self.app = app
+        self.url_prefix = url_prefix or "/swagger"
+        self.spec = spec
+        _dir_path = os.path.dirname(os.path.realpath(__file__))
+        self.dir_path = os.path.abspath(_dir_path + "/ui")
+        if self.app:
+            self.init_app(self.app)
 
-    return json(options)
+    def init_app(self, app) -> None:
+        add_routes(app, self.url_prefix, spec=self.spec, dir_path=self.dir_path)
```

Comment:
Why not put `build_consumer_route_param()`, `build_method()`, `build_paths()`, `build_blueprint()`, `add_routes()` into `Swagger()` class? Is there any use case that I can use all those mathods directly? If not, I would suggest that put all methods into `Swagger()` class.

Reply:
Done